Feat: edit challenge, delete challenge 처리

pongpongie opened the pull request:
## #️⃣ Part

- [x] FE
- [ ] BE

## 📝 작업 내용

challenge edit, giveup, giveupBeforeStart, delete 등을 구현한 페이지와 컨텍스트 등을 추가하였습니다.
`route`에 host만 사용할 수 있는 `editChallenge`와 관련된 route를 추가하였습니다.

pongpongie committed:
feat - add challenge edit api

pongpongie committed:
feat: add ChallengeEditBtn to Setting page

pongpongie committed:
Merge #248 into front/iOS

pongpongie committed:
feat: add EditChallenge page and router

pongpongie committed:
feat: implement giveUpChallenge, deleteChallenge, editChallenge to pages

pongpongie committed:
feat: add challenge delete and edit api

pongpongie committed:
feat: add edit challenge ui

pongpongie committed:
feat: add hostroute and adding mates content card

pongpongie committed:
Merge #251 into front/iOS

pongpongie committed:
fix: make add mates only for hosts

ifindary commented on frontend/src/contexts/ChallengeContext.js:
```diff
@@ -71,6 +71,78 @@ const ChallengeContextProvider = ({ children }) => {
+    } else if (!isGiveUpChallengeLoading && giveUpChallengeError) {
+      console.error(giveUpChallengeError);
+    } else {
+      console.log(response);
```
#250 머지되면 이 부분 에러 처리

```handleError(giveUpChallengeError, giveUpChallengeStatus);```

이걸로 수정해도 괜찮을 것 같습니다! (500 일 경우 에러 페이지로 이동, 그 외는 백엔드에서 넘겨주는 에러 alert으로 노출)

pongpongie replied:
확인했습니다~!

pongpongie replied:
이거 백엔드 api 수정 완료 후에 처리하겠습니다

pongpongie committed:
Merge #249 #253 into front/iOS

pongpongie committed:
Merge branch 'dev-v2' into front/iOS

pongpongie committed:
feat: add handleError to challengeContext

ifindary commented on frontend/src/contexts/ChallengeContext.js:
```diff
@@ -117,9 +212,15 @@ const ChallengeContextProvider = ({ children }) => {
+      //   todayChallengeDataError.message,
+      //   todayChallengeDataError.status,
+      // );
+      // 에러 처리 추가
```
리뷰 다시 요청 와서 확인 중인데, 이 부분도 백엔드 api 수정 후에 이루어질까요?

pongpongie replied:
백에서 처리되고 난 다음에 주석만 해제하면 됩니다!

pongpongie committed:
Merge #253 #255 into front/iOS

pongpongie commented:
This pull request introduces several new features and improvements to the challenge management functionality in the application. The most important changes include adding new routes for editing and managing challenges, updating the challenge services with new methods, and enhancing the challenge context and main page components.

### Routing and Navigation Enhancements:
* [`frontend/src/Router.js`](diffhunk://#diff-276a70bcc661ed270b4a54e6e9294ba76f4defa05c6dada9269a03b900dbc0ddL8-R22): Added `HostRoute` and `EditChallenge` components to the router and nested the `EditChallenge` route under `HostRoute` to ensure only hosts can access it. [[1]](diffhunk://#diff-276a70bcc661ed270b4a54e6e9294ba76f4defa05c6dada9269a03b900dbc0ddL8-R22) [[2]](diffhunk://#diff-276a70bcc661ed270b4a54e6e9294ba76f4defa05c6dada9269a03b900dbc0ddR61-R63)
* [`frontend/src/pages/Auth/HostRoute.js`](diffhunk://#diff-320f520f0e5e9e8fd8adec9d6f2ebcbc8ced6771d8271393c9a47de63c442773R1-R31): Created a new `HostRoute` component to restrict access to certain routes based on the user's role.
* [`frontend/src/pages/Auth/index.js`](diffhunk://#diff-31fc298e0a00694d7735fee38108c0e190bf00b36d20c119454e8752bc4dc80bR5-R7): Exported the newly created `HostRoute` component.

### Challenge Services:
* [`frontend/src/apis/challengeServices.js`](diffhunk://#diff-5bfcddf7e569000308320a13920c5afe749ece7d58a9add2d91613e3819753f9R85-R135): Added new methods for editing, deleting, giving up, and adding mates to challenges. [[1]](diffhunk://#diff-5bfcddf7e569000308320a13920c5afe749ece7d58a9add2d91613e3819753f9R85-R135) [[2]](diffhunk://#diff-5bfcddf7e569000308320a13920c5afe749ece7d58a9add2d91613e3819753f9R171-R175)

### Challenge Context:
* [`frontend/src/contexts/ChallengeContext.js`](diffhunk://#diff-4a5bdbc52a58db9aba7245caabff435543599e7d62cfb62b94ca832778944a81R2-R5): Enhanced the context provider with new state variables and methods for handling challenge edits, giving up challenges, and managing challenge start status. [[1]](diffhunk://#diff-4a5bdbc52a58db9aba7245caabff435543599e7d62cfb62b94ca832778944a81R2-R5) [[2]](diffhunk://#diff-4a5bdbc52a58db9aba7245caabff435543599e7d62cfb62b94ca832778944a81R15-R16) [[3]](diffhunk://#diff-4a5bdbc52a58db9aba7245caabff435543599e7d62cfb62b94ca832778944a81L28-R33) [[4]](diffhunk://#diff-4a5bdbc52a58db9aba7245caabff435543599e7d62cfb62b94ca832778944a81R68-R168) [[5]](diffhunk://#diff-4a5bdbc52a58db9aba7245caabff435543599e7d62cfb62b94ca832778944a81R215-R223) [[6]](diffhunk://#diff-4a5bdbc52a58db9aba7245caabff435543599e7d62cfb62b94ca832778944a81R276-R308)

### Main Page Components:
* [`frontend/src/pages/Main/Main.js`](diffhunk://#diff-b9091aa6c6ad05209675591aae8e447cd48499591873cbcd3d01f1046b6a4856R11): Updated the main page to include a new card for inviting mates to a challenge if the challenge has not started and the user is the host. [[1]](diffhunk://#diff-b9091aa6c6ad05209675591aae8e447cd48499591873cbcd3d01f1046b6a4856R11) [[2]](diffhunk://#diff-b9091aa6c6ad05209675591aae8e447cd48499591873cbcd3d01f1046b6a4856L23-R30) [[3]](diffhunk://#diff-b9091aa6c6ad05209675591aae8e447cd48499591873cbcd3d01f1046b6a4856L52-R54) [[4]](diffhunk://#diff-b9091aa6c6ad05209675591aae8e447cd48499591873cbcd3d01f1046b6a4856R81-R88)
* [`frontend/src/pages/Main/DATA.js`](diffhunk://#diff-7c5c3e0552581c86e5d657cd9dca5cb70d0628341abc934c72d75121ddb9634eR20): Added a new card style for the invite mates feature.

These changes collectively enhance the application's challenge management capabilities, providing more flexibility and control to users, especially hosts.

pongpongie committed:
fix: typo of handleGIveUpBeforeChallenge

pongpongie committed:
Merge branch 'dev-v2' into front/iOS
